shell/platform/windows/compositor_opengl.cc

```diff
@@ -0,0 +1,188 @@
+  } else {
+    return GL_RGBA8;
+  }
+}
+
+}  // namespace
+
+CompositorOpenGL::CompositorOpenGL(FlutterWindowsEngine* engine,
+                                   impeller::ProcTableGLES::Resolver resolver)
+    : engine_(engine), resolver_(resolver) {}
+
+bool CompositorOpenGL::CreateBackingStore(
+    const FlutterBackingStoreConfig& config,
+    FlutterBackingStore* result) {
+  if (!is_initialized_ && !Initialize()) {
+    return false;
+  }
+
+  auto store = std::make_unique<FramebufferBackingStore>();
+
+  gl_->GenTextures(1, &store->texture_id);
+  gl_->GenFramebuffers(1, &store->framebuffer_id);
+
+  gl_->BindFramebuffer(GL_FRAMEBUFFER, store->framebuffer_id);
+
+  gl_->BindTexture(GL_TEXTURE_2D, store->texture_id);
+  gl_->TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+  gl_->TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+  gl_->TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+  gl_->TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+  gl_->TexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, config.size.width,
+                  config.size.height, 0, GL_RGBA, GL_UNSIGNED_BYTE, nullptr);
+  gl_->BindTexture(GL_TEXTURE_2D, 0);
+
+  gl_->FramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0_EXT,
+                            GL_TEXTURE_2D, store->texture_id, 0);
+
+  result->type = kFlutterBackingStoreTypeOpenGL;
+  result->open_gl.type = kFlutterOpenGLTargetTypeFramebuffer;
+  result->open_gl.framebuffer.name = store->framebuffer_id;
+  result->open_gl.framebuffer.target = format_;
+  result->open_gl.framebuffer.user_data = store.release();
+  result->open_gl.framebuffer.destruction_callback = [](void* user_data) {
+    // Backing store destroyed in `CompositorOpenGL::CollectBackingStore`, set
+    // on FlutterCompositor.collect_backing_store_callback during engine start.
+  };
+  return true;
+}
+
+bool CompositorOpenGL::CollectBackingStore(const FlutterBackingStore* store) {
+  FML_DCHECK(is_initialized_);
+  FML_DCHECK(store->type == kFlutterBackingStoreTypeOpenGL);
+  FML_DCHECK(store->open_gl.type == kFlutterOpenGLTargetTypeFramebuffer);
+
+  auto user_data = static_cast<FramebufferBackingStore*>(
+      store->open_gl.framebuffer.user_data);
+
+  gl_->DeleteFramebuffers(1, &user_data->framebuffer_id);
+  gl_->DeleteTextures(1, &user_data->texture_id);
+
+  delete user_data;
+  return true;
+}
+
+bool CompositorOpenGL::Present(const FlutterLayer** layers,
+                               size_t layers_count) {
+  if (!engine_->view()) {
+    return false;
+  }
+
+  // Clear the view if there are no layers to present.
+  if (layers_count == 0) {
+    // Normally the compositor is initialized when the first backing store is
+    // created. However, on an empty frame no backing stores are created and
+    // the present needs to initialize the compositor.
+    if (!is_initialized_ && !Initialize()) {
+      return false;
+    }
+
+    return ClearSurface();
```

Comment:
Clearing the surface requires a context to be made current. In this code path, you only do it after the call to ClearSurface. Is the context is not already previously current, this call will fail.

Reply:
Oh, nevermind. You explicitly make another call in ClearSurface.

Reply:
Yup that's correct. Marking as resolved